selfdrive/camerad/cameras/camera_qcom2.cc

```diff
@@ -540,41 +540,23 @@ static void camera_init(MultiCameraState *multi_cam_state, VisionIpcServer * v,
   s->buf.init(device_id, ctx, s, v, FRAME_BUF_COUNT, rgb_type, yuv_type);
 }
 
-// TODO: refactor this to somewhere nicer, perhaps use in camera_qcom as well
-static int open_v4l_by_name_and_index(const char name[], int index, int flags) {
-  char nbuf[0x100];
-  int v4l_index = 0;
-  int cnt_index = index;
-  while (true) {
-    snprintf(nbuf, sizeof(nbuf), "/sys/class/video4linux/v4l-subdev%d/name", v4l_index);
-    FILE *f = fopen(nbuf, "rb");
-    if (f == NULL) return -1;
-    int len = fread(nbuf, 1, sizeof(nbuf), f);
-    fclose(f);
-
-    // name ends with '\n', remove it
-    if (len < 1) return -1;
-    nbuf[len-1] = '\0';
-
-    if (strcmp(nbuf, name) == 0) {
-      if (cnt_index == 0) {
-        snprintf(nbuf, sizeof(nbuf), "/dev/v4l-subdev%d", v4l_index);
-        LOGD("open %s for %s index %d", nbuf, name, index);
-        return open(nbuf, flags);
-      }
-      cnt_index--;
+int open_v4l_by_name_and_index(const char name[], int index, int flags = O_RDWR | O_NONBLOCK) {
+  for (int v4l_index = 0; /**/; ++v4l_index) {
+    std::string v4l_name = util::read_file(util::string_format("/sys/class/video4linux/v4l-subdev%d/name", v4l_index));
+    if (v4l_name.empty()) return -1;
+    if (v4l_name.compare(0, v4l_name.length() - 1, name) == 0 && index-- == 0) {
```

Comment:
Maybe `v4l_name.find(name) == 0` is clearer? Also keep the comment about there being a `\n`.

Think relying on the short circuiting of `&&` makes the code a little bit harder to read than necessary. Let's keep the nested if from the original PR.

Reply:
much nicer :+1: